### PR TITLE
BUG: Make `random.shuffle` work on 1-D instances of `ndarray` subclasses

### LINKED ISSUE #11442: shuffle on memmap'ed structured array corrupts data
OS: Windows 7 SP1
NumPy: 1.14.5
Python: 3.6.5

## Abstract

Calling `numpy.random.shuffle` on a memory-mapped structured array introduces repeated elements. The first element in the array appears to be most (only?) affected. Repeated calls to `shuffle` will result in an array that just repeats the first element of the original array.

## Code Example

```
import numpy as np
np.random.seed(1)

x = np.array([
    ('aa', 1, 1.1),
    ('bb', 2, 0.7),
    ('cc', 3, 0.5),
    ('xy', 4, 0.3),
    ('zz', 5, 0.2)
], dtype=[
    ('col1', '<U2'),
    ('col2', '<i4'),
    ('col3', '<f4')
])

fp = np.memmap("test.dat", dtype=x.dtype, mode="w+", shape=x.shape)

fp[:] = x
```

before shuffle
 [('aa', 1, 1.1) ('bb', 2, 0.7) ('cc', 3, 0.5) ('xy', 4, 0.3) ('zz', 5, 0.2)]

after shuffle
 [('aa', 1, 1.1) ('bb', 2, 0.7) ('aa', 1, 1.1) ('aa', 1, 1.1) ('xy', 4, 0.3)]

### PR DESCRIPTION
**Pull request by lormuc**
Closes #11442.

### CODE REVIEW

File: numpy/random/tests/test_random.py [hunk 1]
```diff
@@ -467,6 +467,9 @@ def test_shuffle(self):
                      lambda x: [(i, i) for i in x],
                      lambda x: np.asarray([[i, i] for i in x]),
                      lambda x: np.vstack([x, x]).T,
+                     lambda x: (np.asarray([(i, i) for i in x],
+                                           [('', int), ('', int)])
+                                .view(np.recarray)),
```
Comment (tylerjereddy) [inline numpy/random/tests/test_random.py]: you might want to consider labelling with the corresponding issue you are patching: `gh-11442` ; the test below does that

**Review by tylerjereddy — COMMENTED** [review]
Looks pretty positive. To review this I checked that:

1. The example code in the linked issue causes a problem in master but not this PR branch.
2. If I use the updated test file in this PR branch with a build from master branch -- I confirmed that this does fail.

Some notes in passing, though we probably shouldn't "kill the messenger" for a potentially useful fix:

- it is a little hard to read the test failure because of the old-school lambda-style testing used here, but that's just matching how the code has been for some time probably
- we don't currently assess coverage for Cython files, but this is a small enough change anyway & I've done the manual checks above as well

**Comment by tylerjereddy** [comment]
One thing I'm not sure of though is that there's a NEP accepted that effectively blocks bug fixes in random module at some stage for backwards compatibility; not sure if that applies here / yet.

**Comment by rkern** [comment]
This one's fine, IMO. I do recommend a comment next to the new location of the `empty_like()` call. I wouldn't want anyone coming in and "optimizing" it away later.

**Comment by asnasnasn** [comment]
Suggested comments added.

File: numpy/random/mtrand/mtrand.pyx [hunk 2]
```diff
@@ -4836,12 +4836,13 @@ cdef class RandomState:
                     self._shuffle_raw(n, sizeof(npy_intp), stride, x_ptr, buf_ptr)
                 else:
                     self._shuffle_raw(n, itemsize, stride, x_ptr, buf_ptr)
-        elif isinstance(x, np.ndarray) and x.ndim > 1 and x.size:
-            # Multidimensional ndarrays require a bounce buffer.
-            buf = np.empty_like(x[0])
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: I think changing this to `x[0,...]` would do the trick

Comment (asnasnasn) [reply numpy/random/mtrand/mtrand.pyx]: Done.

**Review by jeffyancey — APPROVED** [review]
LGTM, and looks like the two suggestions have been taken.

**Comment by mattip** [comment]
Thanks @asnasnasn

**Pull request merged**